text/0018-generalized-in.md

```diff
@@ -0,0 +1,184 @@
+More formally, the semantics of `a in B` is:
+* if `a` and `B` are both entity references, then `true` if `a` is either equal to `B`, or a descendant of `B` in the hierarchy, else `false`
+    * this is not changed from today's Cedar
+* if `B` is a Cedar set and `a` is not an entity reference, then it's ordinary shallow set membership, i.e., equivalent to `B.contains(a)`
+    * this case is a type error in today's Cedar
+* if `B` is a Cedar set and `a` is an entity reference, then WLOG let `B` be `[B1, B2, B3]`. The semantics of `a in [B1, B2, B3]` is `a op B1 || a op B2 || a op B3` where for each `Bx`, if `Bx` is an entity reference then `op` is `in`, otherwise `op` is `==`
+    * in the case where `B` only contains entity references, this behavior matches today's Cedar
+    * in the case where `B` contains other values, this would be a type error in today's Cedar
+* in all other cases, type error
+    * this is not changed from today's Cedar
+
+Here's one additional example (we'll call it Example 3):
+```
+resource.owner in ["foo", "bar", User::"Group"]
+```
+Under this proposal, this expression is `true` if `resource.owner` is the string `"foo"`, the string `"bar"`, the entity reference `User::"Group"`, or any entity reference which is `in` `User::"Group"`.
+
+Arguably, this is the "intuitive" semantics that users want for `in` -- it blends "deep" `in` for entity references while also providing shallow set-membership for sets.
+
+One final note -- this RFC does not propose any changes to what syntax is legal in the policy scope. And of course, it doesn't change the behavior of any syntax that's currently legal in the policy scope.
+
+## Drawbacks
+
+- The new semantics is more complicated than Cedar's current semantics for `in`. This could lead to additional confusion.
+    - Response: The current semantics also leads to confusion, which this proposal at least partially addresses.
+- The new semantics doesn't provide a way to do shallow set membership with entity references; you will still need `.contains()` for that.
+    - Response: Is there a real-world use case that needs shallow set membership with entity references?
+- The new `in` is nearly redundant with `.contains()`. For ordinary set-membership with non-entities (as in both of the basic examples at the top), users would now have two different ways to do the same thing -- something we have previously said we want to generally avoid if possible.
+    - Response: It seems valuable to have two ways to do set-membership, one with the set on the LHS and the other with the set on the RHS. Cedar already supports "flipped" versions of some operators, which are redundant except for switching their operands -- consider `<` and `>`.
+- A final objection is simply that one of the Alternatives below might be better. See discussion below.
+
+## Alternatives
+
+### Alternative 1: Do nothing (status quo)
+
+The Motivation section above explains why this is undesirable.
```

Comment:
I think doing nothing is the wrong choice, but we could still do something simpler that what's proposed by the RFC to reduce the pain. In particular: we could add a special error message for when someone tries to use `in` in place of `contains`. In the [slack thread](https://cedar-policy.slack.com/archives/C0547KH7R19/p1689253188448809) linked above, the error message reported was:
```
Errors: while evaluating policy policy0, encountered the following error: type error: expected (entity of type any_entity_type), got string
```
It would have been presumably easier to diagnose & fix the issue if we instead reported something to the effect of "type error: `in` expects an entity on the LHS. Use `.contains()` for set membership."

Reply:
+100 to this.  I'll open an issue right now, we can at least do this now while the RFC continues to be under discussion